modules/wallet/database.go

```diff
@@ -38,91 +68,115 @@ func dbDelete(b *bolt.Bucket, key interface{}) error {
 	return b.Delete(encoding.Marshal(key))
 }
 
-// dbPutHistoricClaimStart stores the number of siafunds corresponding to the
-// specified SiafundOutputID. The wallet builds a mapping of output -> value
-// so that it can determine the value of siafund outputs that were not created
-// by the wallet.
+// dbForEach is a helper function for iterating over a bucket and calling fn
+// on each entry. fn must be a function with two parameters. The key/value
+// bytes of each bucket entry will be unmarshalled into the types of fn's
+// parameters.
+func dbForEach(b *bolt.Bucket, fn interface{}) error {
+	// check function type
+	fnVal, fnTyp := reflect.ValueOf(fn), reflect.TypeOf(fn)
+	if fnTyp.Kind() != reflect.Func || fnTyp.NumIn() != 2 {
+		panic("bad fn type: needed func(key, val), got " + fnTyp.String())
+	}
+
+	return b.ForEach(func(keyBytes, valBytes []byte) error {
+		key, val := reflect.New(fnTyp.In(0)), reflect.New(fnTyp.In(1))
+		if err := encoding.Unmarshal(keyBytes, key.Interface()); err != nil {
+			return err
+		} else if err := encoding.Unmarshal(valBytes, val.Interface()); err != nil {
+			return err
+		}
+		fnVal.Call([]reflect.Value{key.Elem(), val.Elem()})
+		return nil
+	})
+}
```

Comment:
The reflection is pretty simple compared to most reflection. It's used in three places: 1) to check that the function has the correct type; 2) to allocate values to decode into; 3) to call the function.

I actually did initially define separate functions; see 0dc9cab for the diff. Though a reflection-based approach may be less readable, defining a separate `ForEach` for each bucket has its downsides as well. The biggest one is that the reader cannot easily confirm that each `ForEach` function is doing the same thing. Even if they look broadly similar, there may be subtle differences. And of course, more code means more places for a bug to creep in.